tests/viscoelastic/zhong_viscoelastic_free_surface.py

```diff
@@ -0,0 +1,188 @@
+# Test case based on simple harmonic loading and unloading problems
+# from `Viscosity of the Earth's Mantle' by Cathles (1975). The specific
+# analytic solution is actually based off of Equation 2 in `On calculating
+# glacial isostatic adjustment', Cathles (2024). The decay time is the viscous
+# relaxation timescale plus the maxwell time, including the elastic buoyancy
+# effects. Note that we are solving a loading problem not an unloading problem.
+# There are three default tests:
+# 1) elastic case limit (dt << maxwell time, 1 step)
+# 2) viscoelastic (dt ~ maxwell time)
+# 3) viscous limit (dt >> maxwell time)
+
+from gadopt import *
+import numpy as np
+import argparse
+OUTPUT = False
+output_directory = "./2d_analytic_zhong_viscoelastic_freesurface/"
+
+parser = argparse.ArgumentParser()
+parser.add_argument("--case", default="viscoelastic", type=str, help="Test case to run: elastic limit (dt << maxwell time, 1 step), viscoelastic (dt ~ maxwell time), viscous limit (dt >> maxwell time) ", required=False)
+args = parser.parse_args()
+
+
+def viscoelastic_model(nx=80, dt_factor=0.1, sim_time="long", shear_modulus=1e11):
+    # Set up geometry:
+    nz = nx  # Number of vertical cells
+    D = 3e6  # length of domain in m
+    L = D/2  # Depth of the domain in m
+    mesh = RectangleMesh(nx, nz, L, D)  # Rectangle mesh generated via firedrake
+    mesh.cartesian = True
+
+    # Squash mesh to refine near top boundary modified from the ocean model
+    # Roms e.g. https://www.myroms.org/wiki/Vertical_S-coordinate
+    mesh.coordinates.dat.data[:, 1] -= D
+    x, z = SpatialCoordinate(mesh)
+    a = Constant(4)
+    b = Constant(0)
+    depth_c = 500.0
+    z_scaled = z / D
+    Cs = (1.-b) * sinh(a*z_scaled) / sinh(a) + b*(tanh(a*(z_scaled + 0.5))/(2*tanh(0.5*a)) - 0.5)
+    Vc = mesh.coordinates.function_space()
+    f = Function(Vc).interpolate(as_vector([x, depth_c*z_scaled + (D - depth_c)*Cs]))
+    mesh.coordinates.assign(f)
+    left_id, right_id, bottom_id, top_id = 1, 2, 3, 4  # Boundary IDs
+
+    # Set up function spaces - currently using P2P1 element pair:
+    V = VectorFunctionSpace(mesh, "CG", 2)  # Displacement function space (vector)
+    W = FunctionSpace(mesh, "CG", 1)  # Pressure function space (scalar)
+    Q = FunctionSpace(mesh, "CG", 2)  # Temperature function space (scalar)
+    Z = MixedFunctionSpace([V, W])  # Mixed function space.
+    TP1 = TensorFunctionSpace(mesh, "DG", 1)
+    R = FunctionSpace(mesh, "R", 0)
+
+    # Function to store the solutions:
+    z = Function(Z)  # a field over the mixed function space Z.
+    u, p = split(z)  # Returns symbolic UFL expression for u and p
+    u_, p_ = z.subfunctions  # Returns individual Function for output
```

Comment:
Can we drop this and use *z.subfunctions or similar when outputting lower down? This would be consistent with our other demos.